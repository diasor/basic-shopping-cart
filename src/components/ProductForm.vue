<template>
	<v-card 
		justify="left"
		elevation="2"
		class="mt-4 pa-10"
		max-width="1000px"
	>

		<v-card-title class="pa-0">Product </v-card-title>
		<v-form
			ref="form"
			v-model="valid"
			lazy-validation
		>
			<v-row justify="space-between">
				<v-col
					cols="12"
					md="6"
				>
					<v-select
						v-model="productType"
						:items="productTypeList"
						:rules="[v => !!v || 'Type is required']"
						label="Type"
						required
						@input="resetProduct()"
					></v-select>

					<v-select
						v-model="productName"
						:items="productNameList"
						:rules="[v => !!v || 'Name is required']"
						label="Name"
						required
						@input="changeProduct()"
					></v-select>

					<v-select
						v-model="productColor"
						:items="productColorList"
						:rules="[v => !!v || 'Color is required']"
						label="Color"
						required
					></v-select>
				</v-col>
				<v-col
					cols="12"
					md="4"
				>
					<v-row justify="center">
						<v-img
							v-if="product.img"
							:src="productImage"
							max-height="300"
							class="mr-10"
							contain
						/>
					</v-row>
				</v-col>
			</v-row>

			<v-row class="mt-6">
				<v-btn
					:disabled="!valid"
					color="success"
					class="mr-4"
					@click="validate"
				>
					Add to Cart
				</v-btn>

				<v-btn
					color="error"
					class="mr-4"
					@click="reset"
				>
					Reset Form
				</v-btn>
			</v-row>
		</v-form>
	</v-card>
</template>
<script>
import { mapGetters, mapActions } from "vuex"

export default {
	name: "ProductForm",

    data: () => ({
		valid: true,
		productType: "",
		productName: "",
		product: {},
		productColor: ""
    }),

	computed: {
		...mapGetters(["productTypes", "allProducts"]),

		productNameList () {
			if (this.productType === [] || this.allProducts.length === 0) {
				return []
			} else {
				return this.allProducts
					.filter(product => product.type === this.productType)
					.map(product => product.name)
			}
		},
		
		productTypeList () {
			return (this.productTypes || [])
		},

		productColorList () {
			if (this.productName === "" || this.allProducts.length === 0) {
				return []
			} else {
				return this.product.colors
			}
		},

		productImage () {
			return (this.product.img) ? this.product.img : ""
		}
	},

    methods: {
		...mapActions(['addToCart']),

		changeProduct () {
			if (this.productName && this.allProducts.length > 0) {
				this.product = this.allProducts.find(product => product.name === this.productName)
			} else {
				this.product = {}
			}
		},

		resetProduct () {
			this.product = {}
			this.productName = ""
			this.productColor = ""
		},

		validate () {
			if (this.$refs.form.validate()) {
				this.addToCart({
					type: this.productType,
					name: this.productName,
					color: this.productColor,
					img: this.product.img
				})
				this.reset()
			}
		},

		reset () {
			this.$refs.form.reset()
		}
    }
  }
</script>
