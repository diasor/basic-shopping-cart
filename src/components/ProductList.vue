<template>
	<v-card
		v-if="myCart.length >0"
		max-width="450"
		class="mx-auto mt-4"
	>
		<v-card-title class="pb-0">My Cart </v-card-title>
		<v-list >
			<v-list-item-group multiple >
				<template v-for="(product, index) in myCart" >
					<v-list-item :key="product.name" > 
						<v-img 
							:src="product.img" 
							class="ma-4 rounded" max-width="70px" max-height="70px" 
							contain
						/>
						<v-list-item-content>
							<v-list-item-title v-html="product.name"></v-list-item-title>
							<v-list-item-subtitle>{{ product.type }} - {{ product.color }}</v-list-item-subtitle>
						</v-list-item-content>
					</v-list-item>
					<v-divider
						v-if="index < myCart.length - 1"
						:key="index"
					/>	
				</template>
			</v-list-item-group>
		</v-list>
	</v-card>
</template>
<script>
import { mapGetters } from "vuex"

export default {
	name: "ProductList",
	
	computed: {
		...mapGetters(["myCart"])
	}
}
</script>
