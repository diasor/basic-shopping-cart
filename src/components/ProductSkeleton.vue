<template>
	<v-container>
		<v-card 
			justify="center"
			elevation="2"
			class="pa-10"
			max-width="900px"
		>

			<v-card-title class="pa-0">Product </v-card-title>
			<v-row justify="space-between">
				<v-col
					cols="12"
					md="4"
					class="pl-0"
				>
					<v-skeleton-loader
						v-bind="attrs"
						class="mt-10"
						type="list-item-three-line, actions"
					></v-skeleton-loader>
				</v-col>
				<v-col
					cols="12"
					md="4"
				>
					<v-skeleton-loader
						v-bind="attrs"
						type="image"
					></v-skeleton-loader>
				</v-col>
			</v-row>
		</v-card>
	</v-container>
</template>
<script>
export default {
	name: "ProductSkelenton",
	
	data: () => ({
		attrs: {
			class: 'mb-6',
			boilerplate: true
		}
	})
}
</script>
