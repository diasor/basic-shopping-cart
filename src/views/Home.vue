<template>
	<div data-testid="home-container">
		<v-row justify="space-between">
			<v-col
				cols="12"
				sm="6"
				md="8"
			>
				<product-skeleton v-if="loadSkeleton"/>
				<product-form v-else/>
			</v-col>
			<v-col
				cols="6"
				md="4"
			>
				<product-list />
			</v-col>
		</v-row>
	</div>
</template>

<script>
import { mapGetters, mapActions } from "vuex"
import ProductForm from "@components/ProductForm.vue"
import ProductSkeleton from "@components/ProductSkeleton.vue"
import ProductList from "@components/ProductList.vue"

export default {
	name: "Home",
	
	components: {
		ProductSkeleton,
		ProductForm,
		ProductList
	},
	
	computed: {
		...mapGetters(["loading"]),

		loadSkeleton () {
			return this.loading
		}
	},

	async created () {
		await this.getProducts()
	},

	methods: {
		...mapActions(['getProducts'])
	}
}
</script>
