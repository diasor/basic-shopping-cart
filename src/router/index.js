import Vue from "vue"
import VueRouter from "vue-router"
import Home from "@views/Home.vue"

Vue.use(VueRouter)

const routes = [
  {
	path: "/error",
	name: 'Error',
	component: () => import(/* webpackChunkName: "about" */ "../views/Error.vue")
  },
  {
    path: "/",
    name: "Home",
    component: Home
  },
  {
	path: "*",
	redirect: '/error'
  }
]

const router = new VueRouter({
	mode: "history",
	routes
})

export default router
