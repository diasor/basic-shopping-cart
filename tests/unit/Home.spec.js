import { shallowMount, createLocalVue } from "@vue/test-utils"
import Vuex from "vuex"
import router from "@router"
import Vuetify from "vuetify"
import Home from "@views/Home.vue"
import ProductForm from "@components/ProductForm.vue"
import ProductSkeleton from "@components/ProductSkeleton.vue"
import ProductList from "@components/ProductList.vue"
import producstStore from "@store/productsStore"
import { mockedProductList } from "@tests/unit/mockData.js"

let localVue
let wrapper
let vuetify
let store

afterEach(() => {
	wrapper.destroy()
})

describe("testing Home Template", () => {
	beforeEach(() => createInstance())
	it("should render the page and match the snapshot", () => {
		expect(wrapper.exists()).toBe(true)
		expect(wrapper).toMatchSnapshot()
	})

	it("should render one container div with the proper name", () =>
		expect(wrapper.find("div[data-testid=\"home-container\"]").exists()).toBe(true))

	it("should render no ProductSkeleton component, one ProductForm component and one ProductList component", () => {
		expect(wrapper.findAllComponents(ProductSkeleton).length).toBe(0)
		expect(wrapper.findAllComponents(ProductForm).length).toBe(1)
		expect(wrapper.findAllComponents(ProductList).length).toBe(1)
	})
})

describe("testing Home Template when the application is loading (show skeleton form)", () => {
	beforeEach(() => createInstance(mockedProductList, true))

	it("should render the page and match the snapshot", () => {
		// tests that the wrapper was loadded properly
		expect(wrapper.exists()).toBe(true)
		expect(wrapper).toMatchSnapshot()
	})

	it("should render one ProductSkeleton component, no ProductForm component and one ProductList component", async() => {
		expect(wrapper.findAllComponents(ProductSkeleton).length).toBe(1)
		expect(wrapper.findAllComponents(ProductForm).length).toBe(0)
		expect(wrapper.findAllComponents(ProductList).length).toBe(1)
	})
})

describe("testing Home Computed properties", () => {
	describe("testing loadSkeleton computed", () => {
		it("should return false if loading is false on the store", () => {
			createInstance()
			expect(wrapper.vm.loadSkeleton).toBe(false)
		})

		it("should return false if loading is false on the store", () => {
			createInstance(mockedProductList, true)
			expect(wrapper.vm.loadSkeleton).toBe(true)
		})
	})
})

/**
 * [createInstance is a function that defines all the boiler plate code to create the Home view.
 * 				 It creates the Home instance, with vuetify, a new store and router.
 * 				 The store is created based on the parameters. This allows the mounting function to
 * 				 simulate different scenarios for testing. ]
* @param {[Array]} 	 dataSet [ represents the list of products to allow testing how the Home instance behaves with products.
							  the default value is an empty array. ]
* @param {[Boolean]} loading [ represents an indicator of whether the application is still loading data.
							  the default value is false. ]
 */
async function createInstance(dataSet = [], loading = false) {
	localVue = createLocalVue()
	localVue.use(Vuex)
	vuetify = new Vuetify()
	
	const state = {
		products : dataSet,
		productTypes: (dataSet.length > 0) ? [ "oven", "fridge", "vacuum" ] : [],
		productNames: (dataSet.length > 0) ? [ "H 6100 BM AM", "F 2412 Vi" ] : [],
		loading,
		cart: []
	}

	const getProducts = jest.fn().mockImplementation(() => Promise.resolve(mockedProductList))
	store = new Vuex.Store({
		state,
		getters: producstStore.getters,
		actions: {
			...producstStore.actions, 
			getProducts
		},
		mutations: producstStore.mutations
	})

	// mount the wrapper
	wrapper = shallowMount(Home, {
		localVue,
		vuetify,
		router,
		store
	})
	await wrapper.vm.$nextTick()
}