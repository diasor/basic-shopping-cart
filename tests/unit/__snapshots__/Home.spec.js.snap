// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing Home Template should render the page and match the snapshot 1`] = `
<div data-testid="home-container">
  <v-row-stub tag="div" justify="space-between">
    <v-col-stub cols="12" sm="6" md="8" tag="div">
      <product-form-stub></product-form-stub>
    </v-col-stub>
    <v-col-stub cols="6" md="4" tag="div">
      <product-list-stub></product-list-stub>
    </v-col-stub>
  </v-row-stub>
</div>
`;

exports[`testing Home Template when the application is loading (show skeleton form) should render the page and match the snapshot 1`] = `
<div data-testid="home-container">
  <v-row-stub tag="div" justify="space-between">
    <v-col-stub cols="12" sm="6" md="8" tag="div">
      <product-skeleton-stub></product-skeleton-stub>
    </v-col-stub>
    <v-col-stub cols="6" md="4" tag="div">
      <product-list-stub></product-list-stub>
    </v-col-stub>
  </v-row-stub>
</div>
`;
